scrypto/src/math/decimal.rs

```diff
@@ -11,506 +10,591 @@ use sbor::rust::vec::Vec;
-from_int!(isize);
-
-impl From<&str> for Decimal {
-    fn from(val: &str) -> Self {
-        Self::from_str(&val).unwrap()
-    }
-}
+                impl<T: Into<$dec>> Mul<T> for $dec {
+                    type Output = $dec;
 
-impl From<String> for Decimal {
-    fn from(val: String) -> Self {
-        Self::from_str(&val).unwrap()
-    }
-}
+                    fn mul(self, other: T) -> Self::Output {
+                        let a = self.0;
+                        let b: $wrapped = TryInto::<$dec>::try_into(other).expect("Overflow").0;
+                        let c: $wrapped = a * b / Self::ONE.0;
+                        $dec(c)
+                    }
+                }
 
-impl From<bool> for Decimal {
-    fn from(val: bool) -> Self {
-        if val {
-            Self::from(1)
-        } else {
-            Self::from(0)
-        }
-    }
-}
+                impl<T: Into<$dec>> Div<T> for $dec {
+                    type Output = $dec;
 
-/// Creates a `Decimal` from literals.
-///
-/// # Example
-/// ```ignore
-/// use scrypto::prelude::*;
-///
-/// let a = dec!(1);
-/// let b = dec!("1.1");
-/// ```
-#[macro_export]
-macro_rules! dec {
-    ($x:literal) => {
-        ::scrypto::math::Decimal::from($x)
-    };
+                    fn div(self, other: T) -> Self::Output {
+                        let a = self.0;
+                        let b: $wrapped = TryInto::<$dec>::try_into(other).expect("Overflow").0;
+                        let c: $wrapped = a * Self::ONE.0 / b;
+                        $dec(c)
+                    }
+                }
 
-    ($base:literal, $shift:literal) => {
-        // Base can be any type that converts into a Decimal, and shift must support
-        // comparison and `-` unary operation, enforced by rustc.
-        {
-            let base = ::scrypto::math::Decimal::from($base);
-            if $shift >= 0 {
-                base * 10i128.pow(u32::try_from($shift).expect("Shift overflow"))
-            } else {
-                base / 10i128.pow(u32::try_from(-$shift).expect("Shift overflow"))
-            }
-        }
-    };
-}
+                impl Neg for $dec {
+                    type Output = $dec;
 
-impl<T: Into<Decimal>> Add<T> for Decimal {
-    type Output = Decimal;
+                    fn neg(self) -> Self::Output {
+                        $dec(-self.0)
+                    }
+                }
 
-    fn add(self, other: T) -> Self::Output {
-        let a = BigInt::from(self.0);
-        let b = BigInt::from(other.into().0);
-        let c = a + b;
-        Self(c.to_i128().expect("overflow"))
-    }
-}
+                impl<T: Into<$dec>> AddAssign<T> for $dec {
+                    fn add_assign(&mut self, other: T) {
+                        self.0 += TryInto::<$dec>::try_into(other).expect("Overflow").0;
+                    }
+                }
 
-impl<T: Into<Decimal>> Sub<T> for Decimal {
-    type Output = Decimal;
+                impl<T: Into<$dec>> SubAssign<T> for $dec {
```

Comment:
@0xOmarA I used `Into` because its shorter than `TryInto` and if a method has an Into trait it sure has a `TryInto` trait as well. And I must use `try_into()` because an `U512` might not fit into a `Decimal`.
So I think it would be less confusing to use `TryInto` trait in the generics, and use your solution as a cleaner one.

Reply:
Fixed in https://github.com/radixdlt/radixdlt-scrypto/commit/ce62163727b49f7c26b2849662cf3b3ff58e6b34

Reply:
Fixed, but please check carefully as the code did not become any more simple.
https://github.com/radixdlt/radixdlt-scrypto/commit/ce62163727b49f7c26b2849662cf3b3ff58e6b34